File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/WelcomeActivity.java
package com.heaton.funnyvote.ui;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.heaton.funnyvote.FirstTimePref;
import com.heaton.funnyvote.MainActivity;
import com.heaton.funnyvote.R;
import com.heaton.funnyvote.data.Injection;
import com.heaton.funnyvote.data.local.AppDatabase;
import com.heaton.funnyvote.data.local.dao.PromotionDao;
import com.heaton.funnyvote.data.user.UserDataRepository;
import com.heaton.funnyvote.database.Promotion;
import com.heaton.funnyvote.ui.introduction.IntroductionActivity;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class WelcomeActivity extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_welcome);
        InitTask task = new InitTask(this);
        task.execute();
    }

    private static class InitTask extends AsyncTask<Void, Void, Void> {

        private WeakReference<Activity> contextWeakReference;
        private SharedPreferences firstTimePref;

        public InitTask(Activity context) {
            contextWeakReference = new WeakReference<>(context);
            firstTimePref = Injection.provideFirstTimePref(contextWeakReference.get());
        }

        @Override
        protected void onPostExecute(Void o) {
            if (contextWeakReference.get() == null) return;
            if (firstTimePref.getBoolean(FirstTimePref.SP_FIRST_INTRODUCTION_PAGE, true)) {
                contextWeakReference.get().startActivity(
                        new Intent(contextWeakReference.get(), IntroductionActivity.class));
            } else {
                contextWeakReference.get().startActivity(
                        new Intent(contextWeakReference.get(), MainActivity.class));
            }
            UserDataRepository userDataRepository = Injection.provideUserRepository(contextWeakReference.get());
            userDataRepository.getUser(true)
                    .subscribeOn(Injection.provideSchedulerProvider().computation())
                    .observeOn(Injection.provideSchedulerProvider().ui())
                    .subscribe(
                            user -> {
                                if (contextWeakReference.get() != null) {
                                    contextWeakReference.get().finish();
                                }
                            },
                            throwable -> {
                                if (contextWeakReference.get() != null) {
                                    contextWeakReference.get().finish();
                                }
                            }
                    );
        }

        @Override
        protected Void doInBackground(Void... voids) {
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            if (firstTimePref.getBoolean(FirstTimePref.SP_FIRST_MOCK_DATA, true)) {
                if (contextWeakReference.get() != null) {
                    PromotionDao promotionDao = AppDatabase.getInstance(contextWeakReference.get().getApplicationContext()).promotionDao();
                    String[] imageURL = contextWeakReference.get().getResources().getStringArray(R.array.imageURL);
                    List<Promotion> promotions = new ArrayList<>();
                    for (int i = 0; i < 1; i++) {
                        Promotion promotion = new Promotion();
                        promotion.setImageURL(imageURL[i % imageURL.length]);
                        promotion.setActionURL("https://play.google.com/store/apps/details?id=com.heaton.funnyvote");
                        promotion.setTitle("title:" + i);
                        promotions.add(promotion);
                    }
                    promotionDao.deleteAll();
                    promotionDao.insertAll(promotions);
                    firstTimePref.edit().putBoolean(FirstTimePref.SP_FIRST_MOCK_DATA, false).apply();
                }
            }
            return null;
        }
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/WelcomeActivity.kt
package com.heaton.funnyvote.ui

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.heaton.funnyvote.FirstTimePref
import com.heaton.funnyvote.MainActivity
import com.heaton.funnyvote.R
import com.heaton.funnyvote.data.Injection
import com.heaton.funnyvote.data.local.AppDatabase
import com.heaton.funnyvote.data.user.UserDataSource
import com.heaton.funnyvote.database.Promotion
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.databinding.ActivityWelcomeBinding
import com.heaton.funnyvote.ui.introduction.IntroductionActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.ArrayList

class WelcomeActivity : AppCompatActivity() {
    private lateinit var binding: ActivityWelcomeBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityWelcomeBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val userDataRepository = Injection.provideUserRepository(applicationContext)

        CoroutineScope(Dispatchers.IO).launch {
            delay(500)
            val firstTimePref = Injection.provideFirstTimePref(applicationContext)

            if (firstTimePref.getBoolean(FirstTimePref.SP_FIRST_MOCK_DATA, true)) {
                val promotionDao = AppDatabase.getInstance(applicationContext).promotionDao()
                val imageURL = resources.getStringArray(R.array.imageURL)
                val promotions = ArrayList<Promotion>()
                for (i in 0..0) {
                    val promotion = Promotion().apply {
                        this.imageURL = imageURL[i % imageURL.size]
                        actionURL = "https://play.google.com/store/apps/details?id=com.heaton.funnyvote"
                        title = "title:$i"
                    }
                    promotions.add(promotion)
                }
                promotionDao.deleteAll()
                promotionDao.insertAll(promotions)
                firstTimePref.edit().putBoolean(FirstTimePref.SP_FIRST_MOCK_DATA, false).apply()
            }

            withContext(Dispatchers.Main) {
                if (firstTimePref.getBoolean(FirstTimePref.SP_FIRST_INTRODUCTION_PAGE, true)) {
                    startActivity(Intent(this@WelcomeActivity, IntroductionActivity::class.java))
                } else {
                    startActivity(Intent(this@WelcomeActivity, MainActivity::class.java))
                }
                userDataRepository.getUser(object : UserDataSource.GetUserCallback {
                    override fun onResponse(user: User) {
                        finish()
                    }

                    override fun onFailure() {
                        finish()
                    }
                }, true)
            }
        }
    }
}
